File: typescript/09-saas-integrator.ts
/**
 * SaaS integrator: you store history, meter usage, watch limits.
 *
 * tonia Pass does not keep SDK chat threads. Billing, seats, and API keys
 * stay in the portal (https://portal.tonia.ca).
 *
 *   export TONIA_API_KEY=tonia_sk_…
 *   npx tsx 09-saas-integrator.ts
 */
import { EntitlementError, RateLimitError, Tonia } from "@tonia-router/sdk";

const client = new Tonia({
  apiKey: process.env.TONIA_API_KEY,
  defaultHeaders: {
    "HTTP-Referer": "https://github.com/tonia-router/sdk-examples",
    "X-Tonia-Title": "sdk-examples",
  },
});

const { data } = await client.models.list();
if (!data[0]) {
  throw new Error("this key has no models; check the profile allowlist in the portal");
}
const model = data[0].id;

type ChatMessage = { role: "user" | "assistant"; content: string };

function assistantText(body: unknown): string {
  const choice = (body as { choices?: { message?: { content?: unknown } }[] })
    ?.choices?.[0];
  const content = choice?.message?.content;
  return typeof content === "string" ? content.trim() : "";
}

function usageOf(body: unknown): {
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
} {
  const usage = (body as { usage?: Record<string, unknown> })?.usage;
  const prompt = usage?.prompt_tokens ?? usage?.input_tokens;
  const completion = usage?.completion_tokens ?? usage?.output_tokens;
  const total = usage?.total_tokens;
  return {
    prompt_tokens: typeof prompt === "number" ? prompt : null,
    completion_tokens: typeof completion === "number" ? completion : null,
    total_tokens: typeof total === "number" ? total : null,
  };
}

// Stand-in for your database. Pass does not store this for the SDK.
const thread: ChatMessage[] = [
  { role: "user", content: "Remember this word: maple. Reply only: ok" },
];

try {
  const first = await client.chat.completions.create({
    model,
    max_tokens: 8,
    messages: thread,
  });
  thread.push({ role: "assistant", content: assistantText(first) || "ok" });
  thread.push({
    role: "user",
    content: "Reply with only the word I asked you to remember.",
  });
  const second = await client.chat.completions.create({
    model,
    max_tokens: 16,
    messages: thread,
  });
  // lastLimits is set only when usage is ≥ 80% of quota/budget.
  console.log(
    JSON.stringify(
      {
        turns: thread.length,
        usage: usageOf(second),
        lastLimits: client.lastLimits
          ? {
              warning: client.lastLimits.warning,
              kind: client.lastLimits.kind,
              remaining: client.lastLimits.remaining,
              periodEndsAt: client.lastLimits.periodEndsAt,
            }
          : null,
      },
      null,
      2,
    ),
  );
} catch (err) {
  if (err instanceof RateLimitError) {
    console.error(
      `Admission limited (${err.reason ?? err.code}). Wait ${err.retryAfterSeconds ?? 1}s, then retry once.`,
    );
    process.exitCode = 1;
  } else if (err instanceof EntitlementError) {
    console.error(
      err.retryable
        ? "Monthly quota reached. Wait Retry-After, or raise the plan in the portal."
        : "Budget exhausted. Do not retry until a Workspace admin adds credit.",
    );
    process.exitCode = 1;
  } else {
    throw err;
  }
}

File: python/09_saas_integrator.py
"""SaaS integrator: you store history, meter usage, watch limits.

tonia Pass does not keep SDK chat threads. Billing, seats, and API keys
stay in the portal (https://portal.tonia.ca).

    export TONIA_API_KEY=tonia_sk_…
    python 09_saas_integrator.py
"""

from __future__ import annotations

import os
import sys

from tonia import EntitlementError, RateLimitError, Tonia


def assistant_text(body: object) -> str:
    if not isinstance(body, dict):
        return ""
    choices = body.get("choices")
    if not isinstance(choices, list) or not choices or not isinstance(choices[0], dict):
        return ""
    message = choices[0].get("message")
    if not isinstance(message, dict):
        return ""
    content = message.get("content")
    return content.strip() if isinstance(content, str) else ""


def usage_of(body: object) -> dict[str, int | None]:
    usage = body.get("usage") if isinstance(body, dict) else None
    if not isinstance(usage, dict):
        return {"prompt_tokens": None, "completion_tokens": None, "total_tokens": None}
    prompt = usage.get("prompt_tokens", usage.get("input_tokens"))
    completion = usage.get("completion_tokens", usage.get("output_tokens"))
    total = usage.get("total_tokens")
    return {
        "prompt_tokens": prompt if isinstance(prompt, int) else None,
        "completion_tokens": completion if isinstance(completion, int) else None,
        "total_tokens": total if isinstance(total, int) else None,
    }


with Tonia(
    api_key=os.environ["TONIA_API_KEY"],
    default_headers={
        "HTTP-Referer": "https://github.com/tonia-router/sdk-examples",
        "X-Tonia-Title": "sdk-examples",
    },
) as client:
    listed = client.models.list()
    ids = [model["id"] for model in listed["data"]]
    if not ids:
        raise SystemExit("empty allowlist — do not guess a model id")
    model = ids[0]

    # Stand-in for your database. Pass does not store this for the SDK.
    thread: list[dict[str, str]] = [
        {"role": "user", "content": "Remember this word: maple. Reply only: ok"},
    ]
    try:
        first = client.chat.completions.create(
            model=model, max_tokens=8, messages=thread
        )
        thread.append({"role": "assistant", "content": assistant_text(first) or "ok"})
        thread.append(
            {
                "role": "user",
                "content": "Reply with only the word I asked you to remember.",
            }
        )
        second = client.chat.completions.create(
            model=model, max_tokens=16, messages=thread
        )
    except RateLimitError as err:
        print(
            f"Admission limited ({err.reason or err.code}). "
            f"Wait {err.retry_after_seconds or 1}s, then retry once.",
            file=sys.stderr,
        )
        raise SystemExit(1) from err
    except EntitlementError as err:
        print(
            "Monthly quota reached. Wait Retry-After, or raise the plan in the portal."
            if err.retryable
            else "Budget exhausted. Do not retry until a Workspace admin adds credit.",
            file=sys.stderr,
        )
        raise SystemExit(1) from err

    # last_limits is set only when usage is ≥ 80% of quota/budget.
    print(
        {
            "turns": len(thread),
            "usage": usage_of(second),
            "last_limits": (
                None
                if client.last_limits is None
                else {
                    "warning": client.last_limits.warning,
                    "kind": client.last_limits.kind,
                    "remaining": client.last_limits.remaining,
                    "period_ends_at": client.last_limits.period_ends_at,
                }
            ),
        }
    )
